### ISSUE TITLE
% symbol is not being displayed through third party plugins on latest beta branch.

### ISSUE THREAD

**Issue by basuritashka** [ValveSoftware/halflife#2611]
Title.

Beta:
%d%% ->[100 ]

Stable version:
%d%% ->[100%]

**Comment by SamVanheer** [comment]
#2531 is resurfacing?

**Comment by afwn90cj93201nixr2e1re** [comment]
@mikela-valve the best way is just replace % and # to unicode analog. But then we can get bug with non displayed symbols on xp.

**Comment by basuritashka** [comment]
> 
> 
> #2531 is resurfacing?

Looks like.

**Comment by mikela-valve** [comment]
Is this only happening on CS or for other games as well? #2531 was a fix to the `SayText` handler exclusively for CS.

This might be some combination of message processing that is causing this. The `SayText` handler by itself does not handle non-string format specifiers, so `%d%%` must be getting handled earlier, replaced with `100%` which then has the % stripped by the `SayText` handler. 

Before I tightened up the `SayText` string handling in #2531 you were allowed to have % characters at the end of the input string even though that should be invalid as a format specifier, so when I fixed that it may have broken this. I'll update the beta later today with a change that lets `SayText` have a bare % at the end of the string again and see if that fixes this.

**Comment by basuritashka** [comment]
@mikela-valve I don't have a way to test it on another game besides CS. If someone else can test it on another game that would be really helpful.

**Comment by basuritashka** [comment]
@mikela-valve was the supposed update to the beta postponed?

**Comment by mikela-valve** [comment]
I'm just waiting on some changes to build now. I had a few more changes to make for #2237.

**Comment by basuritashka** [comment]
Thanks, I will update this asap once the update hits the beta branch.

**Comment by mikela-valve** [comment]
Thanks, it should be updated now, build 8304.

**Comment by oaus** [comment]
@mikela-valve Still not appearing with the new version 8304

**Comment by basuritashka** [comment]
@mikela-valve This is still happening on latest beta branch.

**Comment by SamVanheer** [comment]
Typing `%d%%` into chat prints this:
```
Solokiller :  d %
Solokiller :   d %

```

Note that it prints to console twice and it adds a newline on the second print. It's also inserting another space in the second print.

Normal chat text does seem to work with `%%`, but i haven't tested it with programmatically inserted strings.

@basuritashka How are you sending the `%d%%`? Through `SayText` or some other print function?

**Comment by basuritashka** [comment]
@SamVanheer client_print / client_print_color.

**Comment by SamVanheer** [comment]
Ok so that's still ending up in `SayText` eventually.

**Comment by Arkshine** [comment]
`client_print` uses `TextMsg`. `client_print_color` uses `SayText` if CS, otherwise `TextMsg`.

**Comment by basuritashka** [comment]
I was testing some things with AMXX and these are my results:

![image](https://user-images.githubusercontent.com/33007491/61731512-a4c36f00-ad49-11e9-995a-7c41bdfd685e.png)

![image](https://user-images.githubusercontent.com/33007491/61731509-a2611500-ad49-11e9-80cf-01a5987cfd30.png)

![image](https://user-images.githubusercontent.com/33007491/61731539-b4db4e80-ad49-11e9-9209-3bc3ecaa3aae.png)

**Comment by SamVanheer** [comment]
What happens when you type in `%%` in chat yourself on the same server? Does it also completely remove them then?

**Comment by basuritashka** [comment]
@SamVanheer 
say %%
say %

![image](https://user-images.githubusercontent.com/33007491/61732485-b443b780-ad4b-11e9-8545-c1fe8706e2a3.png)

**Comment by SamVanheer** [comment]
Ok, i did some checking in AMX's code, it looks like this problem is caused by how AMX interacts with the new `SayText` code.

Here are some links:
https://github.com/alliedmodders/amxmodx/blob/1cc7786a4c260ca9ad55fa9fd1c8c415115ead89/amxmodx/amxmodx.cpp#L308-L402
https://github.com/alliedmodders/amxmodx/blob/75cf5f55f9b06e778fc9a87c0bbc84d00be7b805/amxmodx/string.cpp#L38
https://github.com/alliedmodders/amxmodx/blob/2559fcf00aecb19fb6e6f8de846ae37cbcd6d572/amxmodx/CLang.cpp#L269
https://github.com/alliedmodders/amxmodx/blob/5a257a7a42ce34de055d087e5460e2190a5c2ed2/amxmodx/format.cpp#L549-L811

AMX will format your text first before sending it, since `SayText` doesn't support complex format arguments.

Among other things it converts `%%` to `%`:
https://github.com/alliedmodders/amxmodx/blob/5a257a7a42ce34de055d087e5460e2190a5c2ed2/amxmodx/format.cpp#L777-L782

So when you pass `%%` it only sends `%`.

Now we can see that a single one is still supported when entered through chat so the question is why it's being ignored. I see that `Host_Say` has code to check for `%` to prevent passing format inputs to the client, but it doesn't do anything when the `%` is followed by a space or nothing.

I do see code to append `'\n'` at the end of the text, so the client is always going to find one more character at the end of the string, even if it ends with `%`.
This appears to be consistent in AMX as well:
https://github.com/alliedmodders/amxmodx/blob/1cc7786a4c260ca9ad55fa9fd1c8c415115ead89/amxmodx/amxmodx.cpp#L356-L357

I don't immediately see a cause of this problem but this should help @mikela-valve track things down.

I'd suggest trying to send `%%%%` to see if that does print, if so it means that repeated printf formatting is halving the number of them every time until it encounters only one and ignores it. If not, double the number of them until you see something.

If nothing shows up after 64 i think it's safe to say it isn't caused by repeated printf calls.

Now as to why it does work in the public version. In that version of the `SayText` handler it scans the input for `%`, and it replaces the character that comes after with a space if it isn't either the end of the input or a space already:
```
for( int i = 0; i < strlen( szMsg ); ++i )
{
	if( szMsg[ i ] == '%' )
	{
		if( szMsg[ i + 1 ] && szMsg[ i + 1 ] != ' ' )
		{
			szMsg[ i + 1 ] = ' ';
		}
	}
}
```

So `%%` becomes `% ` here. It looks to me like AMX converts `%%` to `%` already, and the client then sees `%\n` and this is somehow ignored.

@Arkshine If you know of any way to log the data being sent using `SayText` we can see exactly what string is being sent to the client both by AMX and the game when using chat so we can figure out which data is causing this result.

I think there might be a Metamod plugin that does this but i don't know if there's anything readily available to run.

**Comment by mikela-valve** [comment]
Ok, all of these explanations are starting to make sense.

So, first, the reason %'s work in in-game chat is that in-game chat is first filtered and written into a temporary string that is then inserted into the chat buffer through a loc token that resolves to that temporary string. That temporary string is not used as a format string so it is not stripped of format specifiers.

The reason that `client_print` and `client_print_color` are not working is that they send strings directly to the `SayText` and `TextMsg` handlers. The first string parameter in the message is used as a format string that is allowed up to four `%s` format specifiers for dynamic replacements. I fixed the validation of that format string in #2531 which is what broke this for the `client_print` functions as it started filtering out single percent signs in the format string as they are not valid format specifiers.

This can be fixed as @SamVanheer suggested by sending `%%%%` through `client_print_color` which should resolve to `%%` in the `SayText` handler and be kept in the input string as a single %. The other fix would be for AMXX to send the formatted string (as it's clearly pre-formatting the string and sending it formatted) as a second string parameter to `SayText` with `"%s"` as the first parameter. Then any text in the second string parameter would be printed verbatim.

**Comment by basuritashka** [comment]
@mikela-valve 

![image](https://user-images.githubusercontent.com/33007491/61735928-03d9b180-ad53-11e9-98c1-d8b2854c4aa6.png)

![image](https://user-images.githubusercontent.com/33007491/61735933-063c0b80-ad53-11e9-9d1f-6aceee936289.png)

![image](https://user-images.githubusercontent.com/33007491/61735904-f6242c00-ad52-11e9-8ae3-7f2844e1408a.png)

**Comment by SamVanheer** [comment]
@Arkshine What's your opinion on this? Do you think this should be fixed in AMX's code? It seems like the best place to put it, since it reduces the amount of changes required to just the implementations for those message sending functions.

**Comment by Arkshine** [comment]
I did not really follow the last changes lately. I remember vaguely I wanted to use `%s` as first argument to fix exploits, but could not because Alfred filtered it (or maybe it was `TextMsg`?). Sorry, I'm a bit out of the loop. Either way, can we now use `%s` for both messages? If so, I don't see any problems to change it in AMXX (at least I don't see potential breakage at first glance).

**Comment by SamVanheer** [comment]
@mikela-valve says you can send two strings, `%s` followed by the actual message to print it without the checks against `%` interfering with printing.

**Comment by mikela-valve** [comment]
@Arkshine Currently what @SamVanheer is describing is for `SayText` only based on the changes I made in #2531. The format string for `SayText` is allowed to have 0-4 plain `%s` format specifiers to match the 0-4 replacement strings you can send in that message.

**Comment by mikela-valve** [comment]
I'm also not sure why @basuritashka's last test printed two %'s. I looked through the AMXX code and everything looked like it worked the way that I thought it would, i.e. `%%%%` would be replaced with `%%` which would then be let through the `SayText` handler and printed as % in chat, but it looks like the entire `%%%%` came through?

**Comment by SamVanheer** [comment]
That's why it would be very helpful to see what was actually sent to the client once the string made it through AMX's code, so we can see what exactly is getting received. I'm not aware of any plugin that actually does this, however.

**Comment by basuritashka** [comment]
I will leave this here if that helps.

![image](https://user-images.githubusercontent.com/33007491/61824705-b3368700-ae2c-11e9-922a-0457722a66aa.png)

![image](https://user-images.githubusercontent.com/33007491/61824709-b5004a80-ae2c-11e9-806f-c84d7c5e96de.png)

**Comment by Arkshine** [comment]
Random test with:

```C
    client_print_color(id, print_team_default, "100%");
    client_print_color(id, print_team_default, "[100%]");
    client_print_color(id, print_team_default, "--");
    client_print_color(id, print_team_default, "100%%");
    client_print_color(id, print_team_default, "[100%%]");
    client_print_color(id, print_team_default, "--");
    client_print_color(id, print_team_default, "100%%%");
    client_print_color(id, print_team_default, "[100%%%]");
    client_print_color(id, print_team_default, "--");
    client_print_color(id, print_team_default, "100%%%%");
    client_print_color(id, print_team_default, "[100%%%%]");
```
What AMXX sends (which is expected):

%_%_ -> %
%_EOS_ -> %
%_unconsumed character_ -> 

![image](https://user-images.githubusercontent.com/360640/61820729-e8a39a80-ae55-11e9-89c7-e72bb9688f7c.png) 

Latest CS stable | AMXX `SayText` _with_ `%s`:
`%s` can't be used.

![image](https://user-images.githubusercontent.com/360640/61823197-92395a80-ae5b-11e9-984b-a26eefd1d2ee.png)

Latest CS stable | AMXX `SayText` _without_ `%s` (default):
Is it expected? I would have expected at least one occurrence with a `%`.
![image](https://user-images.githubusercontent.com/360640/61821917-ad569b00-ae58-11e9-9655-f6f71c037435.png)

Latest CS beta | AMXX `SayText` _without_ `%s` (default):

![image](https://user-images.githubusercontent.com/360640/61817648-4bddfe80-ae4f-11e9-9e4b-dee9f02728d2.png)

Latest CS beta | AMXX `SayText` _with_ `%s`:
Looks like it's the way to go. Once moved to stable, I can make the change in AMXX.
Though the extra lines in console are odd. Doesn't appear in the chat.
![image](https://user-images.githubusercontent.com/360640/61817705-7334cb80-ae4f-11e9-980e-d334c96bf3e8.png)

Highlighted to see the spaces.

I'm quite confused about the current stable version. I've used latest _without_ beta on server/client (3 Apr 2019) and I'm not sure how OP can see `%`. Or did I miss something?

**Comment by basuritashka** [comment]
@Arkshine This is being tested on latest beta branch.

![image](https://user-images.githubusercontent.com/33007491/61824889-23dda380-ae2d-11e9-8d3e-81a2e6957bc8.png)

**Comment by Arkshine** [comment]
I know, I tested on latest stable _and_ beta. I'm just concerned how you see % on the stable version when in my test it doesn't appear. But I guess it doesn't matter since eventually the beta will be moved to the stable branch.

**Comment by basuritashka** [comment]
@Arkshine That's weird. That was tested on several servers on latest stable version and `%` is being printed correctly, I am testing it right now on a hosted server with ReHLDS and it is not being displayed on latest stable version.

Latest stable version:

**Console output:**
![image](https://user-images.githubusercontent.com/33007491/61827589-ebd95f00-ae32-11e9-9f8e-ecfcb609175f.png)

**Code:**
![image](https://user-images.githubusercontent.com/33007491/61827598-f5fb5d80-ae32-11e9-8aa7-d9b9397bfb0b.png)

**In-game:** 

![image](https://user-images.githubusercontent.com/33007491/61827631-0b708780-ae33-11e9-99ec-0a97f270978a.png)

____

**Latest stable version from a friend server.**
~>[I]nz'KiN.- (ID: 783) DISCONNECTED (Mode: Nightmare, Human Sharpshooter Lvl. 8 [100.0% HP], 24/07/2019 - 10:33:00)

**Comment by basuritashka** [comment]
We did some test with @oaus 

Code [1]:
![image](https://user-images.githubusercontent.com/33007491/61833443-53e37180-ae42-11e9-8740-b3bfa1d2f7d1.png)

Output [1]:
![image](https://user-images.githubusercontent.com/33007491/61833457-5c3bac80-ae42-11e9-8cc7-c63fa9ec618a.png)

___

Code [2]:
![image](https://user-images.githubusercontent.com/33007491/61833482-6bbaf580-ae42-11e9-9da9-6309bd683ee4.png)

Output [2]:
![image](https://user-images.githubusercontent.com/33007491/61833486-707fa980-ae42-11e9-9eca-69252b551dcd.png)

___

Code [3]:
![image](https://user-images.githubusercontent.com/33007491/61833499-7c6b6b80-ae42-11e9-8cd9-26e24bbd09ec.png)

Output [3]:
![image](https://user-images.githubusercontent.com/33007491/61833496-7a091180-ae42-11e9-8151-890db8a5c026.png)

___

Code [4]:
![image](https://user-images.githubusercontent.com/33007491/61833541-9efd8480-ae42-11e9-86b0-41997fbb941d.png)

Output [4]:
![image](https://user-images.githubusercontent.com/33007491/61833544-a3c23880-ae42-11e9-8b8d-5a653837236d.png)

**Comment by mikela-valve** [comment]
Ok, I should have this fixed in the next beta. I spent some time debugging through the `SayText` handler and formatting code and noticed that certain format strings were incorrectly being routed to the localization formatter and not the printf formatter which was why `100%%` was being printed as `100%%` and not `100%`.

I also replaced the old format-stripping process in `TextMsg` with the new `%%` and `%s` validation that `SayText` has, so both of those paths should work the same in the next beta.

**Comment by basuritashka** [comment]
Great, I will update this once beta hits the beta branch and will close this if fixed. Thanks for the hard work!

**Comment by basuritashka** [comment]
@mikela-valve There was an update to the Beta Branch for CS 1.6 but looks like everything still the same, or it wasn't included in this one?

**Comment by mikela-valve** [comment]
Build 8308 should have the fix in it for properly handling the format string. Does it not print properly for `%d%%%%` now? It won't be fixed for `%d%%` until AMXX is updated to send the print string as a parameter instead of as the format string.

**Comment by basuritashka** [comment]
Gotcha, good to know about `%d%%`.

`%d%%%%` it's working properly know.

![image](https://user-images.githubusercontent.com/33007491/61845735-7ee5b980-ae72-11e9-9c95-8f1c3dae51e8.png)

![image](https://user-images.githubusercontent.com/33007491/61845736-80af7d00-ae72-11e9-87f1-257a730b632b.png)

**Comment by Arkshine** [comment]
Tried both message _with_ `%s` using https://github.com/ValveSoftware/halflife/issues/2611#issuecomment-514780120

`SayText`
![image](https://user-images.githubusercontent.com/360640/61851725-086baa80-aeb8-11e9-957d-0402765d6d9c.png)

`TextMsg`
![image](https://user-images.githubusercontent.com/360640/61851780-233e1f00-aeb8-11e9-9ceb-d0846bbe063a.png)

Thanks for adding support in `TextMsg`.
Are the extra newlines in `SayText` from the console can be fixed somehow?

**Comment by mikela-valve** [comment]
I’ll look more into that. It must be something that the `SayText` handler is doing that the `TextMsg` handler is not as they’re both routed to the same functions later to handle formatting and printing the messages.

**Comment by Arkshine** [comment]
It looks like some plugins are using game language key such as `client_print(id, print_center, "#Cstrike_TitlesTXT_Not_Enough_Money");` (it uses `TextMsg)`

It seems with `%s`, it's still working with `print_chat` and `print_console` but not with `print_center` (it shows `%s`).
If possible, is something can be done, at least for the sake of consistency, please? that would be great.

It's not that dramatic because there are not much references (and there is alternative solution like `client_printex`), but if AMXX uses `%s`, I would prefer to avoid any breakage and obscure checks to keep compatibility.

**Comment by di57inct** [comment]
If I'm not mistaking, the same problem is present in the stable version too. Once this is properly fixed, can you push it on it too please? Thank you.

**Comment by basuritashka** [comment]
> 
> 
> If I'm not mistaking, the same problem is present in the stable version too. Once this is properly fixed, can you push it on it too please? Thank you.

There is a way to display `%` properly on the latest stable version but seems to not be working at all (I am a bit confused with this method), this should be added in the next official release and if  you are using AMXX you need will need to wait till Arkshine release an update to allow `%%` usage to display `%`, otherwise you will need to use `%%%%`

**Comment by basuritashka** [comment]
I will be closing this because it is fixed. @Arkshine feel free to update this whenever you want when AMXX release an update regarding the usage of `%%`.

**Comment by Arkshine** [comment]
Is `print_center` fixed as well? I can't test it atm. Also, if an update happens, it will be only when CS changes are available on the stable branch.

**Comment by justgo97** [comment]
I guess something should be added to the new client build that allows AMXX to know which build number the client is to allow it to handle both old and new builds correctly ?

**Comment by Arkshine** [comment]
That's not the point. It will work with `print_chat` and `print_console` ; but not with `print_center`. If anything, `print_center` should use the same logic as those both.

**Comment by basuritashka** [comment]
@Arkshine What do you mean with `print_center`? Not displaying `%s` correctly?

Like this?
![image](https://user-images.githubusercontent.com/33007491/62012689-95805f00-b157-11e9-9f35-407abaeaa528.png)

**Comment by Arkshine** [comment]
^ https://github.com/ValveSoftware/halflife/issues/2611#issuecomment-514962589

**Comment by basuritashka** [comment]
If I understood it correctly, do you mean this?

![image](https://user-images.githubusercontent.com/33007491/62012734-163f5b00-b158-11e9-939d-33dc33d309e7.png)

Both `CenterPrint` and `CenterPrintString` are being displayed correctly on latest stable and latest beta branch.

**Comment by Arkshine** [comment]
No, that's not what I mean. If AMXX is patched, you can't send language keys with `print_center` (you can with others `print_`).
You can't test unless you compile AMXX with the patch, which is just:

![image](https://user-images.githubusercontent.com/360640/62013163-1c9cf980-b18f-11e9-8102-3cce2c2f19f1.png)
(actually, you can test by sending directly the message, with `message_begin`, etc.)

**Comment by basuritashka** [comment]
Oh okay, I just misunderstood your comment thinking it was related to actual AMMX builds (1.9 - 1.10).

**Comment by mikela-valve** [comment]
@Arkshine The `print_center` issue will be fixed in the next beta. 

I also checked out why the extraneous newlines were being shown in the console. Are there embedded newlines in the string arguments being passed to `Saytext` in AMXX? The format string  is stripped of newlines but the format specifiers are not being stripped in the version printed to the console. If it is that, I'd rather leave the version printed in the console identical to the input so I'm going to leave it like it is for the moment unless there are reasons to change it to be printed differently.

**Comment by Arkshine** [comment]
@mikela-valve Thanks, much appreciated.

About the newline, it seems historically AMXX has always appended a new line before sending `SaytText` message:
```
msg[len++] = '\n';	// Client expects newline from the server
```
Old code. I can't really say if it was necessary. Is it still needed now? If it's safe, then I don't mind removing it.

**Comment by mikela-valve** [comment]
Yep, I don't think that's necessary anymore. The message string from `say` does get sent to `SayText` with a newline on it but I tested stripping all newlines from the `SayText` string parameters and everything still worked as expected, with fewer newlines in the console as expected.

**Comment by mikela-valve** [comment]
`print_center` should now work as expected in the current beta.

**Comment by di57inct** [comment]
> If I'm not mistaking, the same problem is present in the stable version too. Once this is properly fixed, can you push it on it too please? Thank you.

@mikela-valve

**Comment by Arkshine** [comment]
@mikela-valve 

It seems new lines are actually stripped when using `TextMsg` with

```
WRITE_STRING("%s");
WRITE_STRING(msg);
```

It doesn't happen with `SayText`. 

Is something can be done for `TextMsg`, please?

**Comment by basuritashka** [comment]
@mikela-valve @Arkshine 

Not sure if related but a empty space is created when using **client_print_color** and sometimes a couple of strings will get merged into a single one (look at the "space" text)

![image](https://user-images.githubusercontent.com/33007491/67581320-32827200-f71e-11e9-8e46-4a83607c96e4.png)

**Comment by Arkshine** [comment]
@basuritashka AMXX inserts by default a newline. For `SayText` it's no more needed with latest game version. Just a matter to update AMXX. I've created an issue https://github.com/alliedmodders/amxmodx/issues/766

**Comment by mikela-valve** [comment]
@Arkshine Unfortunately without examining all of the translation strings I can't remove that newline stripping as I'm unsure of what in the game resources might unknowingly have them. It looks like the same single newline stripping is done to all four of the string inputs, could you double the newline on any string args passed through to `TextMsg`?

**Comment by Arkshine** [comment]
Alright. I can double the newline, it's not that dramatic. Thanks for answering.

**Comment by basuritashka** [comment]
@Arkshine Multiple console_print in the same function are not getting displayed correctly, is that related? Because when using more than 1 console_print in the same function the game will just release a single string with everything merged on it. (Look screenshot from latest post from me)

**Comment by Arkshine** [comment]
@mikela-valve Sorry to bother you again. 

Just tested `WRITE_BYTE("%s")` with `TextMsg` under DoD and it doesn't work for `print_center` and `print_chat` (it shows `%s`).  It works for `print_notice` and `print_console`.
The others mods looks fine.

Also, under CS, with `TextMsg` and `print_center`, newlines are stripped. It works on the others mods.

**Comment by mikela-valve** [comment]
It looks like DoD doesn't currently support that through `TextMsg`. A few of the mods have their own implementations of these message handlers. Report that as a new issue and I'll fix it when I have a chance.

For CS, are you referring to stripping of newlines in the format string or the message args?

**Comment by basuritashka** [comment]
@Arkshine print_center is also displaying %s on CS NS version, I know it is not supported here but its worth mentioning.

**Comment by Arkshine** [comment]
@mikela-valve 
About CS, message args. 

E.g.
```
WRITE_STRING("%s");
WRITE_STRING("Hello \n World");
```

**Comment by Arkshine** [comment]
> print_center is also displaying %s on CS NS version, I know it is not supported here but its worth mentioning.

I guess I will have no choice to check mods and use message args only for the official ones. :weary:

**Comment by mikela-valve** [comment]
@Arkshine It looks like `TextMsg` only treats carriage returns as valid line splitting characters. See if converting \n’s and \f’s to \r’s works.

Were ordinary newlines working for `TextMsg` before when you were passing them in the primary string?

**Comment by Arkshine** [comment]
@mikela-valve 

`\r` works for CS.
`\r` and `\n` seems to work for the others games.

For CS, with or without format string, `\n` is stripped. I can't test on previous build.
But I'm pretty sure it was working before because looking at some plugins, they use `\n`.
It can be fixed in AMXX, but if possible, can `\n` be allowed again?

**Comment by mikela-valve** [comment]
@Arkshine I've been looking back through the changes for `TextMsg` and I don't see how `\n` ever worked for `HUD_PRINTCENTER`. There's explicit checks that allow `\r` to be used as a line separator and to skip `\n` so it looks to me like it should always have worked this way. Additionally the checks are via integer values, not characters (i.e. 10 rather than `\n`) so it shouldn't be the case that CS is interpreting those characters differently. 

It looks like all of the other modes of `TextMsg` either properly interpret `\r` and `\n` or auto-convert all `\r` to `\n`. Is it possible the plugins you checked were using a different HUD mode?

**Comment by Arkshine** [comment]
@mikela-valve 

Honestly, I can't remember. I know plugins using `\n` for sure.
Here some result from an old plugins dump:
![image](https://user-images.githubusercontent.com/360640/67721616-460c3200-f9d7-11e9-84a1-61485e40f811.png)
`print_center` = 4 and `^n` = `\n`

But a fast search on our forum, and I see reference in comments where `^n` doesn't work and  `^r` needs to be used (2017). So, you may be right.

Either way, what about converting `\n` to `r` then? The point is having the same behavior for all mods.

That's said, I've already pushed a fix in AMXX to convert them for CS only. Ideally, I would prefer a fix in the game for the sake of consistency if possible, but it's honestly not a big deal.

**Comment by djearthquake** [comment]
>  so it shouldn't be the case that CS is interpreting those characters differently.

 The Linux CS I reported does not have 'console' Carriage Return phenomenon while OP4 does. Matching Amxx 5260 ⌘773. Buffalo, **_New York._**
